sim-administration/simmanager/src/main/kotlin/org/ostelco/simcards/admin/PeriodicProvisioningTask.kt

```diff
@@ -72,19 +68,27 @@ class PreallocateProfilesTask(
                         }
                     }
 
-    private fun batchPreprovisionSimProfiles(hlrEntry: HssEntry,
-                                             profile: String,
+    private fun batchPreprovisionSimProfiles(hssEntry: HssEntry,
+                                             simProfileName: String,
                                              profileStats: SimProfileKeyStatistics) {
         val noOfProfilesToActuallyAllocate =
                 Math.min(maxNoOfProfileToAllocate.toLong(), profileStats.noOfUnallocatedEntries)
 
         for (i in 1..noOfProfilesToActuallyAllocate) {
-            simInventoryDAO.findNextNonProvisionedSimProfileForHss(hssId = hlrEntry.id, profile = profile)
+
+            // XXX This is all well, if allocation doesn't fail, but if it fails, it will try the
+            //     same profile forever, so something else should be done e.g. setting the
+            //     state of the profile to "provision failed" or something of that nature, so that it
+            //     is possible to move along.   This is an error in the logic of this code.
+            simInventoryDAO.findNextNonProvisionedSimProfileForHss(hssId = hssEntry.id, profile = simProfileName)
                     .flatMap { simEntry ->
-                        preProvisionSimProfile(hlrEntry, simEntry)
+                        preProvisionSimProfile(hssEntry, simEntry)
                                 .mapLeft {
                                     logger.error("Preallocation of SIM ICCID {} failed with error: {}}",
                                             simEntry.iccid, it.description)
+
+                                    // XXX This looks weird, get review on it (and the code above)
+                                    simInventoryDAO.setSmDpPlusState(simEntry.id!!, SmDpPlusState.ALLOCATION_FAILED)
```

Comment:
The '!!' will cause a KotlinNullException if the variable is not set. Maybe some pre check before using the variable? Or better, ensuring that `id` is always set if `simEntry` exists.

Reply:
The !! is necessary since the class permits null ID values, so that instances can be created before the datbase ID is known. When reading from the database this will never happen (obviously), but the type system isn't informed about this.    

In this case an explicit test for null values of the simEntry.id field  could make the "!!" redundant, but I actually don't know how to do that using Arrow.  Any suggestions?

Reply:
I added some comments essentially describing the above then I'll let it pass.   Getting rid of "!!" is a bigger issue anyway. However, this is also an issue where my understanding of control flow in arrow gets in the way of fixing an issue.

Reply:
I found a way (a kludgy if statement :-)